server/src/main/java/org/elasticsearch/index/mapper/SourceValueFetcher.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.search.lookup.SourceLookup;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * An implementation of {@link ValueFetcher} that knows how to extract values
+ * from the document source. Most standard field mappers will use this class
+ * to implement value fetching.
+ */
+public abstract class SourceValueFetcher implements ValueFetcher {
+    private final Set<String> sourcePaths;
+    private final @Nullable Object nullValue;
+    private final boolean parsesArrayValue;
+
+    public SourceValueFetcher(String fieldName, MapperService mapperService, boolean parsesArrayValue) {
+        this(fieldName, mapperService, parsesArrayValue, null);
+    }
+
+    /**
+     * @param fieldName The name of the field.
+     * @param parsesArrayValue Whether the fetcher handles array values during document parsing.
+     * @param nullValue A optional substitute value if the _source value is 'null'.
+     */
+    public SourceValueFetcher(String fieldName, MapperService mapperService, boolean parsesArrayValue, Object nullValue) {
+        this.sourcePaths = mapperService.sourcePath(fieldName);
+        this.nullValue = nullValue;
+        this.parsesArrayValue = parsesArrayValue;
+    }
+
+    @Override
+    public List<Object> fetchValues(SourceLookup lookup) {
+        List<Object> values = new ArrayList<>();
+        for (String path : sourcePaths) {
+            Object sourceValue = lookup.extractValue(path, nullValue);
+            if (sourceValue == null) {
+                return List.of();
+            }
+
+            if (parsesArrayValue) {
```

Comment:
I think it'd be cleaner if this were handled by a differently named subclass. I know we have the boolean in the mapper, but it just feels funny to pass a boolean swapping out the guts of the implementation.

Reply:
It definitely seems cleaner in the abstract, but this approach matches the logic `DocumentParser` where we check `FieldMapper#parsesArrayValue` to decide what value to pass. I also think using `parsesArrayValue` in both places encourages field mappers to have a consistent approach between `parseCreateField` and `parseSourceValue`.

I agree this is not ideal, I really wish there were a solid way to tie together the index-time parsing and value fetching logic...